File: server/src/main/scala/producer/message_generator/data_generators/string/StringGenerator.scala
package producer.message_generator.data_generators.string

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import producer.message_generator.json_generator.JsonGenerator

case class StringGenerator(
    generator: FixedStringGenerator | JsonGenerator
)

object StringGenerator:
    def fromPb(v: pb.StringGenerator): StringGenerator =
        val generator = v.generator match
            case pb.StringGenerator.Generator.GeneratorFixedString(v) => FixedStringGenerator.fromPb(v)
            case pb.StringGenerator.Generator.GeneratorFromJsonString(v)        => JsonGenerator.fromPb(v)
            case _                                                    => throw new Exception("Invalid string generator type")

        StringGenerator(
            generator = generator
        )

    def toPb(v: StringGenerator): pb.StringGenerator =
        val generator = v.generator match
            case v: FixedStringGenerator => pb.StringGenerator.Generator.GeneratorFixedString(FixedStringGenerator.toPb(v))
            case v: JsonGenerator        => pb.StringGenerator.Generator.GeneratorFromJsonString(JsonGenerator.toPb(v))

        pb.StringGenerator(
            generator = generator
        )

File: ui/components/ui/ProducerSession/producer-task/message-generator/data-generators/string-generator/string-generator.ts
import * as pb from '../../../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { JsonGenerator, jsonGeneratorFromPb, jsonGeneratorToPb } from '../json-generator/json-generator';
import { FixedStringGenerator, fixedStringGeneratorFromPb, fixedStringGeneratorToPb } from './fixed-string-generator';

export type StringGenerator = {
  type: 'string-generator',
  generator: FixedStringGenerator | JsonGenerator
};

export function stringGeneratorFromPb(v: pb.StringGenerator): StringGenerator {
  let generator: StringGenerator['generator'];
  switch (v.getGeneratorCase()) {
    case pb.StringGenerator.GeneratorCase.GENERATOR_FIXED_STRING:
      generator = fixedStringGeneratorFromPb(v.getGeneratorFixedString()!);
      break;
    case pb.StringGenerator.GeneratorCase.GENERATOR_FROM_JSON_STRING:
      generator = jsonGeneratorFromPb(v.getGeneratorFromJsonString()!);
      break;
    default:
      throw new Error(`Unknown string generator: ${v.getGeneratorCase()}`);
  }

  return { type: 'string-generator', generator };
}

export function stringGeneratorToPb(v: StringGenerator): pb.StringGenerator {
  const message = new pb.StringGenerator();
  if ('string' in v.generator) {
    message.setGeneratorFixedString(fixedStringGeneratorToPb(v.generator));
  } else {
    message.setGeneratorFromJsonString(jsonGeneratorToPb(v.generator));
  }

  return message;
}
